PokeAlarm/Utilities/PvpUtils.py

```diff
@@ -179,6 +188,36 @@ def get_pvp_info(monster_id, form_id, atk, de, sta, lvl):
                 lvl, ultra_level)
 
     return (float("{0:.2f}".format(great_rating)), great_id, great_cp,
-            great_level, great_candy, great_stardust,
-            float("{0:.2f}".format(ultra_rating)), ultra_id,
-            ultra_cp, ultra_level, ultra_candy, ultra_stardust)
+            great_level, great_candy, great_stardust, great_rank,
+            float("{0:.2f}".format(ultra_rating)), ultra_id, ultra_cp,
+            ultra_level, ultra_candy, ultra_stardust, ultra_rank)
+
+
+def get_pvp_rank(monster_id, form_id, maxcp):
+    if not hasattr(get_pvp_rank, 'info'):
+        get_pvp_rank.info = {}
+    if not hasattr(get_pvp_rank.info, f'cp{maxcp}'):
+        get_pvp_rank.info[f'cp{maxcp}'] = {}
+        file_ = utils.get_path(f'data/rankings-{maxcp}.json')
+        with open(file_, 'r') as f:
+            j = json.load(f)
+            f.close()
+            rank_number = 1
+            for id_ in j:
+                mon_and_form_array = re.split(
+                    r' \(|\)', id_.get('speciesName'))
+                if len(mon_and_form_array) == 1:
+                    mon_and_form_array.append("Normal")
+                elif len(mon_and_form_array) == 3:
+                    mon_and_form_array.pop(2)
+                mon_and_form_str = '_'.join(mon_and_form_array)
+                get_pvp_rank.info[f'cp{maxcp}'][mon_and_form_str] = rank_number
+                rank_number += 1
+
+    loc = locale.Locale('en')
+    mon_name = loc.get_english_pokemon_name(monster_id)
+    form_name = loc.get_english_form_name(monster_id, form_id)
+    form_name = form_name.replace("Alola", "Alolan")
```

Comment:
I had another look at the ranking file, wish it had dex IDs instead

We should probably use the `speciesId` field from the ranking file instead, it's too easy to come by abnormalities when doing string manipulation on `speciesName` like this.

the `speciesId` field is more predictable since you can just lowercase the mon name + the form name with some underscores
To support this more, special characters seem to just be removed with exception of the `.` in Mr. Mime
Some examples:
```
"speciesId": "muk_alolan",
"speciesName": "Muk (Alolan)",
"rating": 622,
...
"speciesId": "sirfetchd",
"speciesName": "Sirfetch'd",
"rating": 690,
...
"speciesId": "mr_mime",
"speciesName": "Mr. Mime",
"rating": 335,
```
I know some other monsters have special characters in their name (ex `é`), didn't look up any of those; would you please find some and see how they work in this file?

I'd imagine this would be very easy with some kind of [slugify](https://github.com/un33k/python-slugify) ability here, and just change the seperator to `_`

Reply:
The problem when you use `speciesId` is that you are unable to know if the monster has a special form whereas the parenthesis in `speciesName` bring you this information. When there is no parenthesis, I set the form name to `Normal` to match with the given monster. In our data, everything which doesn't have a special form is set to Normal.

The truth is there is no easier/clean way to proceed. We have a mon id and a form id in input. If they used them in rankings datas, it would have greatly improved the ease of programming this function.

Reply:
I changed the way the rank is calculated to use `speciesId` from the rankings file and `proto` from the masterfile. Checking some special mon names, this is a more robust way to use both fields.

Checked some random monsters and special ones like Mr. Mime and Sirfetch'd.